modules/nextflow/src/main/groovy/nextflow/processor/TaskProcessor.groovy

```diff
@@ -1942,7 +1942,7 @@ class TaskProcessor {
             if( item instanceof Path || coerceToPath ) {
                 def path = normalizeToPath(item)
                 def target = executor.isForeignFile(path) ? foreignFiles.addToForeign(path) : path
-                def holder = new FileHolder(target)
+                def holder = new FileHolder(path, target)
                 files << holder
```

Comment:
Checking this, and I'm bit concerned on the different logic of the two constructors 

```
    FileHolder( Path inputFile ) {
        assert inputFile
        this.sourceObj = inputFile
        this.storePath = real(inputFile)
        this.stageName = norm(inputFile.getFileName())
    }

    FileHolder( def origin, Path path ) {
        assert origin != null
        assert path != null

        this.sourceObj = origin
        this.storePath = path
        this.stageName = norm(path.getFileName())
    }
```

Above all, note the different logic for `storePath`. I believe this will change the handle on symlink above all. 

If the goal is to bring the source path with could try using the first constructor and adding a method to update the source path

Reply:
I understand your concern.

Would you rather this get implemented as follows?

```suggestion
                def holder = new FileHolder(target)
                holder.setSourceObj(origin)
```

Reply:
Any reason why we can't resolve symlinks in the two-arg constructor?

```groovy
    FileHolder( def origin, Path path ) {
        this.sourceObj = origin
        this.storePath = real(path)
        this.stageName = norm(path.getFileName())
    }
```